Added support for genre, narrator, series and part

sandreas opened the pull request:
This PR add support for `genre`, `narrator`, `series` and `part` in some(!) supported formats (e.g. `mp3 and m4a) and is the first part of a fix for #3044. Since `androidx/media` did not support `movement` and `part`, it hat to be a custom binary reader.

It might need some polish, but in my tests it did work like expected.

Changes:
- Updated database schema and FTS
- Integrated supported fields in default metadata extraction 
- Added `MetaVisitor` for `m4a` (custom implementation, due to the lack of support in androidx/media

sandreas commented on core/scanner/src/main/kotlin/voice/core/scanner/MediaAnalyzer.kt:
```diff
@@ -181,10 +183,26 @@ internal class MediaAnalyzer(
   ) {
     val value = entry.values.first()
     when (entry.id) {
+      // "TPUB" -> Builder.publisher = value
```
Reverted.

sandreas replied:
Done.

sandreas commented on core/scanner/src/main/kotlin/voice/core/scanner/MediaAnalyzer.kt:
```diff
@@ -66,6 +66,8 @@ internal class MediaAnalyzer(
                 is ChapterFrame -> visitChapter(entry, builder)
                 is VorbisComment -> visitVorbis(entry, builder)
                 is MdtaMetadataEntry -> visitMdta(entry, builder)
+                // is InternalFrame -> visitInternalFrame(entry, builder) // id=----, domain=com.apple.iTunes, description=AUDIBLE_ASIN,text=08154711
```
Reverted.

sandreas commented on core/data/api/src/main/kotlin/voice/core/data/BookContent.kt:
```diff
@@ -24,6 +24,14 @@ public data class BookContent(
+  @ColumnInfo(defaultValue = "")
+  val series: String?,
+  @ColumnInfo(defaultValue = "")
+  val part: String?,
```
> I would generally not add the part right now at all.
> Part is not really useful so far so I'd omit it from everything.

While this might be correct for the usual suspects (`Harry Potter` with only 7 parts) I still see huge value in searching for part numbers for more unusual book series with > 30 parts (e.g. `The wheel of time`), especially childrens books with > 100 parts (you're german, so you know `Benjamin Blümchen`, `Bibi Blocksberg`, `Die 3 Fragezeichen`, etc.).

Use case:
- I often remember the number of the last book my daughter listened to and she would like to continue with the next book
- We stopped at number 123
- Now if I search for `Benjamin Blümchen`, I get like 150 audio books and have to scroll the whole list
- If I could search for `Benjamin Blümchen 124` ideally it would show 1 result or at least the first match would be number 124

What do you think? You can't resist a `Benjamin Blümchen` use case, can you? ;-)

sandreas commented on core/scanner/src/main/kotlin/voice/core/scanner/mp4/visitor/MetaVisitor.kt:
```diff
@@ -0,0 +1,162 @@
+    resetParserValuesToDefaults()
+  }
+
+  fun resetParserValuesToDefaults() {
```
Reverted.

sandreas commented on core/scanner/src/main/kotlin/voice/core/scanner/mp4/visitor/MetaVisitor.kt:
```diff
@@ -0,0 +1,162 @@
+        break
+      }
+      val atomSize = buffer.readInt()
+      val atomName = buffer.readString(4, Charsets.ISO_8859_1) // this charset is required to correctly read `©`
```
Reverted.

sandreas commented on features/bookOverview/src/main/kotlin/voice/features/bookOverview/overview/BookOverviewItemViewState.kt:
```diff
@@ -15,6 +15,10 @@ data class BookOverviewItemViewState(
   val progress: Float,
   val id: BookId,
   val remainingTime: String,
+  val genre: String? = "",
```
Done.

sandreas commented:
Thank you for taking the time and providing such a detailed review.

> What do you think about moving these changes into a separate PR and doing the id3 implementation here and then we'll look at it in isolation?

Yeah I've kind of already expected this... I wasn't totally happy with it either. So I reverted the last commit containing all the `m4a` based changes (including the audio files which I took from https://github.com/Zeugma440/atldotnet/) - now everything except the dead code should be fine. I also removed the new properties from ViewState. I hope I did not miss something, it's pretty early in the morning ;-)

sandreas commented:
Good news! Seems that I convinced the `androidx/media` devs to integrate support for at least MovementName and MovementIndex: https://github.com/androidx/media/issues/2754#issuecomment-3248447869

That will take time to be released, but probably we can either skip the whole mp4 byte dance or replace it with something better as soon as it gets released.

PaulWoitaschek commented on core/data/api/src/main/kotlin/voice/core/data/BookContent.kt:
```diff
@@ -24,6 +24,14 @@ public data class BookContent(
   val cover: File?,
   @ColumnInfo(defaultValue = "0")
   val gain: Float,
+  @ColumnInfo(defaultValue = "")
```
Yep please remove it. From an object oriented perspective I very much like it if it's really nullable

sandreas commented:
Ok, I removed the default values, updated the formatting and added the new schema (58.json). Let me know if here is anything left to do.

sandreas commented:
Fixed some missing constructor parameters. I wonder why these weren't reported by Android Studio in the first place or inspection. However, let's hope these are also fixed now.

sandreas commented:
> Did you try to run the tests, as laid out in the project documentation?

Yeah but since the first command 

```
./gradlew voiceUnitTests
```

failed with:

```
FAILURE: Build failed with an exception.

* What went wrong:
Task 'voiceUnitTests' not found in root project 'voice' and its subprojects.
``

I stopped reading - my bad, if I had read further or more carefully, I would have seen the

```
Some candidates are: 'voiceUnitTest'.
```

I try again, but the docs should probably also be updated ;) I'll report back, if everything runs smoothly.

sandreas commented:
Ok, I ran:

```bash
./gradlew formatKotlin
# commit + push

./gradlew voiceUnitTest
./gradlew voiceDeviceGithubDebugAndroidTest
./gradlew lintKotlin
```

All successful. My bad, sorry for wasting your time.
